src/main/java/roomescape/service/ReservationTimeService.java

```diff
@@ -0,0 +1,115 @@
+        this.reservationTimeRepository = reservationTimeRepository;
+        this.reservationRepository = reservationRepository;
+    }
+
+    public List<ReservationTimeResponseDto> getAllTimes() {
+        List<ReservationTime> reservationTimes = reservationTimeRepository.findAll();
+
+        return reservationTimes.stream().map(ReservationTimeResponseDto::from).toList();
+    }
+
+    public ReservationTimeResponseDto saveTime(ReservationTimeRegisterDto reservationTimeRegisterDto) {
+        validateReservationTime(reservationTimeRegisterDto);
+
+        ReservationTime reservationTime = reservationTimeRegisterDto.convertToTime();
+        ReservationTime savedReservationTime = reservationTimeRepository.save(reservationTime);
+
+        return new ReservationTimeResponseDto(savedReservationTime.getId(), savedReservationTime.getStartAt());
+    }
+
+    public void deleteTime(Long id) {
+        try {
+            reservationTimeRepository.deleteById(id);
+        } catch (DataIntegrityViolationException e) {
+            throw new ResourceInUseException("삭제하고자 하는 시각에 예약된 정보가 있습니다.");
+        }
+    }
+
+    public List<AvailableReservationTimeResponseDto> getAvailableTimes(String date, Long themeId) {
+        List<Reservation> reservations = getReservationsBy(date, themeId);
+        List<ReservationTime> reservationTimes = reservationTimeRepository.findAll();
+        Set<ReservationTime> nonDuplicatedReservationTimes = getReservationTimes(reservations);
+
+        reservationTimes.removeAll(nonDuplicatedReservationTimes);
+
+        return getAvailableReservationTimes(reservationTimes, nonDuplicatedReservationTimes);
+    }
+
+    private List<Reservation> getReservationsBy(String date, Long themeId) {
+        LocalDate parsedDate = LocalDate.parse(date);
+        return reservationRepository.findByTheme_IdAndDate(themeId, parsedDate);
+    }
+
+    private Set<ReservationTime> getReservationTimes(List<Reservation> reservations) {
+        return reservations.stream()
+                .map(Reservation::getReservationTime)
+                .collect(Collectors.toSet());
+    }
+
+    private List<AvailableReservationTimeResponseDto> getAvailableReservationTimes(
+            List<ReservationTime> reservationTimes,
+            Set<ReservationTime> nonDuplicatedReservationTimes) {
+        List<AvailableReservationTimeResponseDto> availableReservationTimes = getAvailableReservationTimes(
+                reservationTimes);
+        List<AvailableReservationTimeResponseDto> nonAvailableReservationTimes = getAvailableReservationTimes(
+                nonDuplicatedReservationTimes);
+        availableReservationTimes.addAll(nonAvailableReservationTimes);
+        return availableReservationTimes;
+    }
+
+    private List<AvailableReservationTimeResponseDto> getAvailableReservationTimes(
+            List<ReservationTime> reservationTimes) {
+        return new java.util.ArrayList<>(reservationTimes.stream()
```

Comment:
```suggestion
        return reservationTimes.stream()
```
`ArrayList`로 감쌀 필요가 없어보이네요.

Reply:
```java
private List<AvailableReservationTimeResponseDto> getAvailableReservationTimes(
            List<ReservationTime> reservationTimes,
            Set<ReservationTime> nonDuplicatedReservationTimes) {
        List<AvailableReservationTimeResponseDto> availableReservationTimes = getAvailableReservationTimes(
                reservationTimes);
        List<AvailableReservationTimeResponseDto> nonAvailableReservationTimes = getAvailableReservationTimes(
                nonDuplicatedReservationTimes);
        availableReservationTimes.addAll(nonAvailableReservationTimes);
        return availableReservationTimes;
    }
```

현재 위의 메서드의 다음 부분으로 인해 ArrayList 로 감싸지 않으면 예외가 발생합니다..!

```java
 availableReservationTimes.addAll(nonAvailableReservationTimes);
```

Reply:
```java
        return Stream.of(availableReservationTimes, nonAvailableReservationTimes)
                .flatMap(Collection::stream)
                .toList();
```
가변성을 유지하기보다는 어떻게 불변성을 유지할 수 있을지 고민해보시면 좋을 것 같아요.

Reply:
해당 부분 반영해두었습니다! 가변성을 유지하고 있는 코드라는 생각을 못했는데.. 감사합니다 오리!

Reply:
[refactor: 불변성을 유지하도록 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/226/commits/db6bdca46425c26d2321b128e5851d5b70ef180e)